### PR TITLE
Convert beatmap discussions header to typescript

### PR DESCRIPTION
**Pull request by notbakaneko**
- [x] #10176

### CODE REVIEW

File: resources/js/beatmap-discussions/header.tsx [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+  discussionStarters: UserJson[];
+  events: BeatmapsetEventJson[];
+  mode: DiscussionPage;
+  selectedUserId: number | null;
+  users: Partial<Record<number, UserJson>>;
+}
+
+const statTypes: Filter[] = ['mine', 'mapperNotes', 'resolved', 'pending', 'praises', 'deleted', 'total'];
+
+export class Header extends React.Component<Props> {
```
Comment (nanaya) [inline resources/js/beatmap-discussions/header.tsx:54]: PureComponent?

File: resources/js/beatmap-discussions/header.tsx [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+        <div className='osu-page osu-page--small'>{this.renderHeaderBottom()}</div>
+      </>
+    );
+  }
+
+  private readonly createLink = (beatmap: BeatmapJson) => makeUrl({ beatmap });
+
+  private readonly getCount = (beatmap: BeatmapExtendedJson) => beatmap.deleted_at == null ? this.props.currentDiscussions.countsByBeatmap[beatmap.id] : undefined;
+
+  private onClickMode = (event: React.MouseEvent<HTMLAnchorElement>, mode: GameMode) => {
```
Comment (nanaya) [inline resources/js/beatmap-discussions/header.tsx:84]: missing readonly (also the other)

File: resources/js/beatmap-discussions/header.tsx [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+              </div>
+              <BeatmapBasicStats beatmap={this.props.currentBeatmap} beatmapset={this.props.beatmapset} />
+            </div>
+          </div>
+        </div>
+      </div>
+    );
+  }
+
+  private renderStats() {
```
Comment (nanaya) [inline resources/js/beatmap-discussions/header.tsx:210]: not sure if this is worth the extra function

File: resources/js/beatmap-discussions/header.tsx [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+          <div className={`${bn}__count`}>
+            {total}
+          </div>
+        </div>
+        <div className={`${bn}__line`} />
+      </a>
+    );
+  };
+
+  private readonly setFilter = (event: React.SyntheticEvent<HTMLElement>) => {
```
[the remark this answers is not in the document]
Comment (notbakaneko) [reply resources/js/beatmap-discussions/header.tsx:255]: doesn't really do anything with the MouseEvent-specific parts, though 🤷

**Review by nanaya — APPROVED** [review]

**Pull request merged**